File: bot/bot.cpp
#include "bot.h"
#include <algorithm>
#include <iostream>
#include <QTextStream>
#include "logger.h"
#include <algorithm>
#include <cstdlib>
#include <string>


Bot::Bot(Language *language, int width, int height, QObject* parent) :
    WIDTH(width),
    HEIGHT(height),
    Player(parent),
    language(language)
{
    MOVES.push_back(qMakePair(0, 1));
    MOVES.push_back(qMakePair(0, -1));
    MOVES.push_back(qMakePair(1, 0));
    MOVES.push_back(qMakePair(-1, 0));
}

int Bot::maximalLength(QVector<Word> variants) {

    int maxx = 0;
    for(int i = 0; i<variants.size(); ++i) {
        if (variants[i].possibleWord.length() > maxx)
            maxx = variants[i].possibleWord.length();
    }

    return maxx;

}

QVector<Word> Bot::trulyAllowedWord(QVector<Word> variants, QVector<QString> notAllowedWords) {
    QVector<Word> trulyAllowedVariants;
    for (int i = 0; i<variants.size(); ++i) {
       bool possible = true;
       for(int j = 0; j<notAllowedWords.size(); ++j) {
           if (variants[i].possibleWord == notAllowedWords[j]) {
               possible = false;
           }
       }
       if (possible) {
           trulyAllowedVariants.push_back(variants[i]);
       }
    }

    return trulyAllowedVariants;

}

int Bot::easyIndexWord(QVector<Word> variants) {

    int sz = variants.size() - 1;
    int index = rand()%sz;
    return index;

}

int Bot::mediumIndexWord(QVector<Word> variants) {

    int sz = variants.size() - 1;

    int val = 0;
    for (int i = 0; i < sz; ++i) {
        val += variants[i].possibleWord.length()*variants[i].possibleWord.length();
    }

    int ran = rand()%val;

    int cur = 0;
    for(int i = 0; i < variants.size(); ++i) {
        cur += variants[i].possibleWord.length()*variants[i].possibleWord.length();
        if (cur >= ran)
            return i;
    }

}

int Bot::hardIndexWord(QVector<Word> variants) {

    int sz = variants.size();
    if (sz == 0)
        return -1;

    int maxx = -1;
    for(int i = 0; i<sz; ++i) {
        Logger l;
        l.printLog(DEBUG, variants[i].possibleWord);
        if (variants[i].possibleWord.length() > maxx) {
            maxx = variants[i].possibleWord.length();
        }
    }

    Logger l;
    l.printLog(DEBUG, "meaning of maxx");
    l.printLog(DEBUG, maxx);

    std::vector<Word> maximalVariants;

    for(int i = 0; i<sz; ++i) {
        if (variants[i].possibleWord.length() == maxx) {
            maximalVariants.push_back(variants[i]);
        }
    }

    int newSz = maximalVariants.size();
    if (newSz == 0)
        return -1;

    int indexNewVector = rand()%newSz;

    for(int index = 0; index < variants.size(); ++index) {
        if (variants[index].possibleWord == maximalVariants[indexNewVector].possibleWord) {
            return index;
        }
    }

}

int Bot::hardestIndexWord(QVector<Word> variants, QVector<QVector<QChar> > symbols, QVector<QString> notAllowedWords) {

    int bestIndex = -1;
    int maxx = -111111;
    for(int index = 0; index < variants.size(); ++index) {
        int usedX;
        int usedY;
        QChar c;
        for (int i = 0; i < variants[index].coordinates.size(); ++i) {
            int x = variants[index].coordinates[i].first;
            int y = variants[index].coordinates[i].second;
            if (symbols[x][y] == QChar('-')) {
                usedX = x;
                usedY = y;
                c = variants[index].possibleWord[i];
            }
        notAllowedWords.push_back(variants[index].possibleWord);
        int del1 = variants[index].possibleWord.length();
        symbols[usedX][usedY] = c;
        QVector<Word> newVariants = possibleVariants(symbols);
        symbols[usedX][usedY] = QChar('-');
        notAllowedWords.back();
        QVector<Word> newGoodVariants = trulyAllowedWord(newVariants, notAllowedWords);
        int del2 = maximalLength(newGoodVariants);
        if (del1 - del2 > maxx) {
            maxx = del1 - del2;
            bestIndex = index;
        }
    }
    }
    return bestIndex;
}


void Bot::connectToDictionary(QObject* dictionary) {
    connect(this, SIGNAL(getDictionary()), dictionary, SLOT(sendDictionary()));
    emit getDictionary();
}



void Bot::beginStep() {
    runProcess();
}
/*
*  This slot realizes the move of Bot and interacts with error signals from other classes
*
*  Strategy
*      @author Aliaksei Semchankau
*  Interface
*      @author Pavel Akhtyamov
*/
void Bot::runProcess() {


    QVector<QString> notAllowedWords;

    emit showBoard();

    QVector<QVector<QChar> > symbols(HEIGHT + 2, QVector<QChar>(WIDTH + 2, QChar('#')));
    for (int i = 1; i <= HEIGHT; ++i) {
        for (int j = 1; j <= WIDTH; ++j) {
            symbols[i][j] = board[i-1][j-1];
        }
    }



    QVector<Word> variants = possibleVariants(symbols);

    emit showBoard();

    Logger l;



    isCommited = false;  

    int id = -1;

    if (level == EASY) {
        id = easyIndexWord(variants);
    }

    if (level == MEDIUM) {
        id = mediumIndexWord(variants);
    }

    if (level == HARD) {
        id = hardIndexWord(variants);
    }

    if (level == HARDEST) {
        id = hardestIndexWord(variants, symbols, notAllowedWords);
    }


     if (id == -1) {
         emit dontMakeMove();
     }

    int usedX;
    int usedY;
    QChar c;
    while(!tempCommited) {

        l.printLog(DEBUG, variants[id].possibleWord);
        for (int i = 0; i < variants[id].coordinates.size(); ++i) {
            int x = variants[id].coordinates[i].first - 1;
            int y = variants[id].coordinates[i].second - 1;
            if (board[x][y] == QChar('-')) {
                usedX = x;
                usedY = y;
                c = variants[id].possibleWord[i];
            }
        }

        l.printLog(DEBUG, usedX);
        l.printLog(DEBUG, usedY);
        l.printLog(DEBUG, c);
        QPair<QPair<int,int>, QChar> res(QPair<int,int>(usedX,usedY), c);
        emit chooseLetter(res);

        for (int i = 0; i < variants[id].coordinates.size(); ++i) {
            int x = variants[id].coordinates[i].first - 1;
            int y = variants[id].coordinates[i].second - 1;
            QPair<int,int> coords(x, y);
            l.printLog(DEBUG, "push");
            l.printLog(DEBUG, x);
                l.printLog(DEBUG, y);
            emit pushLetter(coords);
        }

        emit commitWord();

        if (tempCommited) {
            tempCommited = false;
            return;
        }
        //++id;
        QVector<Word> :: iterator it = variants.begin() + id;
        variants.erase(it);

        if (level == EASY) {
            id = easyIndexWord(variants);
        }

        if (level == HARD) {
            id = hardIndexWord(variants);
        }

        if (level == MEDIUM) {
            id = mediumIndexWord(variants);
        }


        if (level == HARDEST) {
            id = hardestIndexWord(variants, symbols, notAllowedWords);
        }


         if (id == -1) {
             emit dontMakeMove();
         }


    }


}

void Bot::setLevel(int difficulty) {
    level = difficulty;
}

bool Bot::notBelong(QVector<QString> notAllowedWords, QString &checkIn) {
    for (int i = 0; i<notAllowedWords.size(); ++i)
        if (notAllowedWords[i] == checkIn)
            return false;

    return true;
}

void Bot::dfs(QVector<QVector<QChar> >  table, QVector<Word> &words,
              int x, int y, int curPosition, QVector<QVector<bool> > curUsed,
              QString &curString,
              QVector<QPair<int, int> > &curCoords, bool usedEmpty)
{
    curUsed[x][y] = true;

    if (borVocabulary.borVertices[curPosition].isLeaf() && usedEmpty)  // Если мы пришли в такую позицию, что слово уже готово. Проверяется использованность хоть какой-то пустой клетки.
    {
        curCoords.push_back(qMakePair(x, y));
        curString += table[x][y];

        words.push_back(Word(curString, curCoords));

        curString = curString.left((int)curString.size() - 1);
        curCoords.pop_back();
    }

    if (table[x][y] == QChar('-') && !usedEmpty)  // Если мы пришли в пустую клетку во время работы дфса, то ставим туда всевозможные буквы и пытаемся продлить слово.
    {

    QVector <QChar> letters = language->getList();
        for (int i=0; i < letters.size(); i++) {
            QChar cc = QChar(letters[i]);
            table[x][y] = cc;
            if (borVocabulary.borVertices[curPosition].findChildren(cc) != -1)
                dfs(table, words, x, y, borVocabulary.borVertices[curPosition].findChildren(cc), curUsed, curString, curCoords, true);
        }
        table[x][y] = QChar('-');
    }
    else
        for (size_t i = 0; i<MOVES.size(); ++i)  // Иначе мы пытаемся пройти в одном из четырех направлений и продлить слово.
        {
        int dx = MOVES[i].first;
        int dy = MOVES[i].second;

        int xx = x + dx;
        int yy = y + dy;

        if (table[xx][yy] == QChar('#'))
            continue;

        if (table[xx][yy] == QChar('-') && usedEmpty)
            continue;

        if (curUsed[xx][yy])
            continue;

        curCoords.push_back(qMakePair(x, y));
        curString += table[x][y];

        if (table[xx][yy] != QChar('-'))
        {
            if (borVocabulary.borVertices[curPosition].findChildren(table[xx][yy]) != NOT_FOUND)
                dfs(table, words, xx, yy, borVocabulary.borVertices[curPosition].findChildren(table[xx][yy]), curUsed, curString, curCoords, usedEmpty);
        }
        else
        {
            dfs(table, words, xx, yy, curPosition, curUsed, curString, curCoords, usedEmpty);
        }

        curString = curString.left((int)curString.size() - 1);
        //curString.pop_back();
        curCoords.pop_back();
    }

    curUsed[x][y] = false;
}

QVector<Word> Bot::possibleVariants (QVector<QVector<QChar> >& table)
{
    int N, M;
    N = table.size() - 2;
    M = table[0].size() - 2;
    QVector<Word> words;

    for (int i = 1; i <= N; ++i)
        for (int j = 1; j <= M; ++j)
        {														// Пытаемся запустить дфс из всех клеток таблицы.
        QVector<QVector<bool> > curUsed(N + 2, QVector<bool>(M + 2, false));
        QVector<QPair<int, int> > curCoords;
        QString curString;

        int curPosition = 0;
        if (table[i][j] != QChar('-') && borVocabulary.borVertices[curPosition].findChildren(table[i][j]) != -1)
            dfs(table, words, i, j, borVocabulary.borVertices[curPosition].findChildren(table[i][j]), curUsed, curString, curCoords, false);
        else
            if (table[i][j] == QChar('-'))
                dfs(table, words, i, j, PRE_VERTEX, curUsed, curString, curCoords, false);

        }

    QVector<Word> corrects;

    for (int i = 0; i < words.size(); ++i)
        corrects.push_back(words[i]);

    return corrects;
}


void Bot::setupDictionary(QVector<QString> words) {
    Logger l;
    l.printLog(DEBUG, tr("Get words"));
    for (int i = 0; i < words.size(); ++i) {
        borVocabulary.addWord(words[i]);
    }
}

File: python-src/bot/Bot.py
from random import Random
from Dictionary import Dictionary
from Letter import CellLetter, Coordinates
from Player import Player
from bot.Bor import Bor, PRE_VERTEX, NOT_FOUND
from bot.Word import Word
from lang.Language import Language

__author__ = 'akhtyamovpavel'

from PySide import QtCore

EASY = 1
MEDIUM = 2
HARD = 3
HARDEST = 100500

DEFAULT_WIDTH = 5
DEFAULT_HEIGHT = 5


class Bot(Player):
    __get_dictionary__ = QtCore.Signal()
    __get_used_words__ = QtCore.Signal()
    send_letter = QtCore.Signal(CellLetter)
    send_word = QtCore.Signal(str)

    @QtCore.Slot()
    def begin_step(self):
        self.run_process()

    @QtCore.Slot()
    def setup_dictionary(self, words):
        print("Get words")
        for word in words:
            self.__bor_vocabulary__.add_word(word)

    @QtCore.Slot()
    def get_used_dictionary(self, used_words):
        self.__not_allowed_words__ = used_words

    def __init__(self, language: Language, width=DEFAULT_WIDTH, height=DEFAULT_HEIGHT):
        super(Bot, self).__init__()
        self.__language__ = language
        self.__width__ = width
        self.__height__ = height
        self.__moves__ = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        self.__level__ = EASY #default value
        self.__bor_vocabulary__ = Bor()
        self.__not_allowed_words__ = set()

    def maximal_length(self, variants):
        if len(variants) == 0:
            return 0
        return max(len(variant.__possible_word__) for variant in variants)

    def allowed_variants(self, variants, not_allowed_word):
        return [variant for variant in variants if variant not in self.__not_allowed_words__ and variant != not_allowed_word]


    def easy_index_word(self, variants):
        print("easy")
        size = len(variants)
        random = Random()
        random_number = int(size * random.random())
        return min(random_number, size - 1)

    def medium_index_word(self, variants):
        print("medium")
        size = len(variants)
        val = 0
        for i in range(size):
            val += len(variants[i].__possible_word__) * len(variants[i].__possible_word__)

        random = Random()
        random_number = int(random.random() * val)

        cur = 0
        cnt = 0
        for variant in variants:
            cur += len(variant.__possible_word__) * len(variant.__possible_word__)
            if cur > random_number:
                return cnt
            cnt += 1

        return size - 1


    def hard_index_word(self, variants):
        print("hard")
        size = len(variants)
        if (size == 0):
            return -1

        max_len = self.maximal_length(variants)
        maximal_variants = [variant for variant in variants if len(variant.__possible_word__) == max_len]
        maximal_size = len(maximal_variants)
        if (maximal_size == 0):
            return -1

        random = Random()
        random_number = min(maximal_size - 1, int(maximal_size * random.random()))
        for cnt in range(size):
            if variants[cnt] == maximal_variants[random_number]:
                return cnt

        return -1

    def hardest_index_word(self, variants, symbols):
        best_index = -1
        max_score = -self.__width__ * self.__height__

        max_lens = {len(variant.__possible_word__) for variant in variants}
        for i in range(3):
            if (len(max_lens) > 0):
                min_len = max(max_lens)
            max_lens.remove(min_len)
        maximal_variants = [variant for variant in variants if len(variant.__possible_word__) >= min_len]

        for index in range(len(maximal_variants)):
            for i in range(len(maximal_variants[index].__coordinates__)):
                x = maximal_variants[index].__coordinates__[i].x
                y = maximal_variants[index].__coordinates__[i].y
                if symbols[x][y] == '-':
                    used_x = x
                    used_y = y
                    c = maximal_variants[index].__possible_word__[i]

            symbols[used_x][used_y] = c
            new_variants = self.possible_variants(symbols)
            symbols[used_x][used_y] = '-'
            new_allowed_variants = self.allowed_variants(new_variants, maximal_variants[index].__possible_word__)

            bot_score = len(maximal_variants[index].__possible_word__)
            player_score = self.maximal_length(new_allowed_variants)
            if (bot_score - player_score > max_score):
                max_score = bot_score - player_score
                best_index = index

        if (best_index == -1):
            return -1

        for cnt in range(len(variants)):
            if variants[cnt] == maximal_variants[best_index]:
                return cnt
        return -1


    def set_level(self, difficulty):
        self.__level__ = difficulty

    def connect_to_dictionary(self, dictionary: Dictionary):
        self.__get_dictionary__.connect(dictionary.send_dictionary)

    def get_dictionary(self):
        self.__get_dictionary__.emit()

    def connect_to_used_dictionary(self, dictionary: Dictionary):
        self.__get_used_words__.connect(dictionary.send_used_words)

    def begin_step(self):
        self.run_process()

    def run_process(self):
        self.show_board.emit()
        self.__get_used_words__.emit()
        print(self.__not_allowed_words__)

        symbols = [['#' for i in range(self.__width__ + 2)] for j in range(self.__height__ + 2)]

        for i in range(1, self.__height__ + 1):
            for j in range(1, self.__width__ + 1):
                symbols[i][j] = self.__board__[i - 1][j - 1]

        variants = self.possible_variants(symbols)
        is_committed = False

        id = -1

        if self.__level__ == EASY:
            id = self.easy_index_word(variants)
        elif self.__level__ == MEDIUM:
            id = self.medium_index_word(variants)
        elif self.__level__ == HARD:
            id = self.hard_index_word(variants)
        else:
            id = self.hardest_index_word(variants, symbols[:])

        if id == -1:
            self.don_t_make_move.emit()

        while not self.__temp_committed__:
            cnt = 0
            for coordinate in variants[id].__coordinates__:
                x = coordinate.x - 1
                y = coordinate.y - 1
                if self.__board__[x][y] == '-':
                    print(cnt)
                    used_x = x
                    used_y = y
                    c = variants[id].__possible_word__[cnt]
                cnt += 1

            res = CellLetter(used_x, used_y, c)
            self.choose_letter.emit(res)

            for coordinate in variants[id].__coordinates__:
                x = coordinate.x - 1
                y = coordinate.y - 1
                coordinates = Coordinates(x, y)
                self.push_letter.emit(coordinates)

            self.commit_word.emit()

            if self.__temp_committed__:
                self.__temp_committed__ = False
                return
            variants.pop(id)

            if self.__level__ == EASY:
                id = self.easy_index_word(variants)
            elif self.__level__ == MEDIUM:
                id = self.medium_index_word(variants)
            elif self.__level__ == HARD:
                id = self.hard_index_word(variants)
            else:
                id = self.hardest_index_word(variants, symbols[:])

            if id == -1:
                self.don_t_make_move.emit()



    def not_belong(self, not_allowed_words, check_in):
        for word in not_allowed_words:
            if word == check_in:
                return False
        return True

    def dfs(self, table, words, x, y, cur_position, cur_used, cur_string, cur_coords, used_empty):
        cur_used[x][y] = True
        if self.__bor_vocabulary__.bor_vertices[cur_position].is_leaf() and used_empty:
            cur_coords.append(Coordinates(x, y))
            cur_string += table[x][y]
            words.append(Word(cur_string, cur_coords))
            cur_string = cur_string[:-1]
            cur_coords.pop()

        if table[x][y] == '-' and not used_empty:
            letters = self.__language__.get_list()
            for letter in letters:
                table[x][y] = letter

                if self.__bor_vocabulary__.bor_vertices[cur_position].find_children(letter) != -1:
                    self.dfs(table, words, x, y,
                             self.__bor_vocabulary__.bor_vertices[cur_position].find_children(letter),
                             cur_used, cur_string, cur_coords, True)
            table[x][y] = '-'
        else:
            for (dx, dy) in self.__moves__:
                xx = x + dx
                yy = y + dy
                if table[xx][yy] == '#':
                    continue
                if table[xx][yy] == '-' and used_empty:
                    continue
                if cur_used[xx][yy]:
                    continue

                cur_coords.append(Coordinates(x, y))
                cur_string += table[x][y]

                if table[xx][yy] != '-':
                    if self.__bor_vocabulary__.bor_vertices[cur_position].find_children(table[xx][yy]) != NOT_FOUND:
                        self.dfs(table, words, xx, yy,
                            self.__bor_vocabulary__.bor_vertices[cur_position].find_children(table[xx][yy]),
                            cur_used, cur_string, cur_coords, used_empty)
                else:
                    self.dfs(table, words, xx, yy, cur_position, cur_used, cur_string, cur_coords, used_empty)

                cur_string = cur_string[:-1]
                cur_coords.pop()

        cur_used[x][y] = False


    def possible_variants(self, table):
        N = len(table) - 2
        M = len(table[0]) - 2
        words = list()

        for i in range(1, N + 1):
            for j in range(1, M + 1):
                cur_used = [[False for k in range(M + 2)] for l in range(N + 2)]
                cur_coordinates = list()
                cur_string = str()
                cur_position = 0
                if table[i][j] != '-' and self.__bor_vocabulary__.bor_vertices[cur_position].find_children(table[i][j]) != -1:
                    self.dfs(table, words, i, j,
                             self.__bor_vocabulary__.bor_vertices[cur_position].find_children(table[i][j]),
                             cur_used, cur_string, cur_coordinates, False)
                elif table[i][j] == '-':
                        self.dfs(table, words, i, j, PRE_VERTEX, cur_used, cur_string, cur_coordinates, False)

        corrects = list()

        for word in words:
            corrects.append(word)

        return corrects



